internal/app/decision_maker.go

```diff
@@ -87,68 +91,81 @@ func (cs *currentState) makePlan(s *state) *plan {
 
 // decide makes a decision about what commands (actions) need to be executed
 // to make a release section of the desired state come true.
-func (cs *currentState) decide(r *release, n *namespace, p *plan, c *chartInfo, settings config) {
+func (cs *currentState) decide(r *release, n *namespace, p *plan, c *chartInfo, settings config, retries int) error {
+	prefix := "Release [ " + r.Name + " ] in namespace [ " + r.Namespace + " ]"
 	// check for presence in defined targets or groups
 	if !r.isConsideredToRun() {
 		if !settings.SkipIgnoredApps {
-			p.addDecision("Release [ "+r.Name+" ] ignored", r.Priority, ignored)
+			p.addDecision(prefix+" ignored", r.Priority, ignored)
 		}
-		return
+		return nil
 	}
 
 	if r.isProtected(cs, n) {
-		p.addDecision("Release [ "+r.Name+" ] in namespace [ "+r.Namespace+" ] is PROTECTED. Operations are not allowed on this release until "+
+		p.addDecision(prefix+" is PROTECTED. Operations are not allowed on this release until "+
 			"protection is removed.", r.Priority, noop)
-		return
+		return nil
 	}
 
 	if flags.destroy {
 		if ok := cs.releaseExists(r, ""); ok {
-			p.addDecision("Release [ "+r.Name+" ] will be DELETED (destroy flag enabled).", r.Priority, remove)
+			p.addDecision(prefix+" will be DELETED (destroy flag enabled).", r.Priority, remove)
 			r.uninstall(p)
 		}
-		return
+		return nil
 	}
 
 	if !r.Enabled {
 		if ok := cs.releaseExists(r, ""); ok {
-			p.addDecision("Release [ "+r.Name+" ] is desired to be DELETED.", r.Priority, remove)
+			p.addDecision(prefix+" is desired to be DELETED.", r.Priority, remove)
 			r.uninstall(p)
 		} else {
-			p.addDecision("Release [ "+r.Name+" ] disabled", r.Priority, noop)
+			p.addDecision(prefix+"is disabled", r.Priority, noop)
 		}
-		return
+		return nil
 	}
 
 	switch cs.releaseStatus(r) {
 	case helmStatusDeployed:
 		if err := cs.inspectUpgradeScenario(r, p, c); err != nil { // upgrade or move
-			log.Fatal(err.Error())
+			return err
 		}
 
 	case helmStatusUninstalled:
 		r.rollback(cs, p) // rollback
 
 	case helmStatusFailed:
-		p.addDecision("Release [ "+r.Name+" ] in namespace [ "+r.Namespace+" ] is in FAILED state. Upgrade is scheduled!", r.Priority, change)
+		p.addDecision(prefix+" is in FAILED state. Upgrade is scheduled!", r.Priority, change)
 		r.upgrade(p)
-		return
+		return nil
 
 	case helmStatusPendingInstall, helmStatusPendingUpgrade, helmStatusPendingRollback, helmStatusUninstalling:
-		log.Fatal("Release [ " + r.Name + " ] in namespace [ " + r.Namespace + " ] is in a pending (install/upgrade/rollback or uninstalling) state. " +
-			"This means application is being operated on outside of this Helmsman invocation's scope." +
-			"Exiting, as this may cause issues when continuing...")
+		if settings.SkipPendingApps {
```

Comment:
as you are using `if settings.SkipPendingApps` condition here, is it worth passing another argument `retries` to decide()? I think you can leave decide() as it was when it comes to arguments and just use `settings.pendingAppRetries` here for retries calculation

Reply:
You mean modify the value from the settings? I think the problem with that is that it is global and we call the decision function for each release and they would after each other. 

Or maybe I'm not getting what you are proposing.

Reply:
no, I just meant "why do we even have to pass `retries` to decide()" when we can just use settings.pendingAppRetries in that logic's part responsible for doing a retry when pending state is encountered :-)

Reply:
It's using recursion, so I don't have to duplicate the logic to check the state and decide, it starts by passing the max number from `settings.pendingAppRetries` but this is passed by copy so each instance of `decide` has its own value, then the `reties` argument is decremented and `decide` calls itself.

We run `decide` for each release, sometimes even in parallel, if I was to use the global `settings.pendingAppRetries` and decrement that it would only work for one of the instances. We pass it as an argument, so each instance has its own value and we can keep track of how many reties that instance has attempted.

I can do this without recursion with a loop inside the `decide` function but I'd have to duplicate some code from the top of the function into that loop, that's why I decided to go with recursion.

Reply:
I see. Ok then. :-)